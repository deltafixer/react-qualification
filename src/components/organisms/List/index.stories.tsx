import React from "react";
import List, { ListProps } from "./";
import { IGitHubRepo } from "../../../models/IGitHubRepo";
import { action } from "@storybook/addon-actions";

export default {
  component: List,
  title: "List",
  // excludes dummy objects
  excludeStories: /.*Data$/,
};

const propsData: ListProps = {
  items: [
    {
      name: "Item 1",
      owner: {
        avatar_url: "https://avatars3.githubusercontent.com/u/69631?v=4",
      },
      created_at: new Date("2019, 04, 10").toString(),
      updated_at: new Date("2020, 01, 10").toString(),
    },
    {
      name: "Item 2",
      owner: {
        avatar_url: "https://avatars3.githubusercontent.com/u/69631?v=4",
      },
      created_at: new Date("2019, 04, 10").toString(),
      updated_at: new Date("2020, 01, 12").toString(),
    },
    {
      name: "Item 3",
      owner: {
        avatar_url: "https://avatars3.githubusercontent.com/u/69631?v=4",
      },
      created_at: new Date("2019, 04, 10").toString(),
      updated_at: new Date("2020, 01, 15").toString(),
    },
  ] as IGitHubRepo[],
  itemFilter: "",
  resultCount: () => {},
};

export const actionsData = {
  resultCount: action("resultCount"),
};

// hardcoding styled div wrapper because storybooks theme is dark, app is white
// would make it smarter if it wasn't a test project (not to go deep in this when no need)

export const DefaultAndSorted = () => (
  <div style={{ background: "white", width: "100%", margin: "auto" }}>
    <List {...(propsData as ListProps)} {...actionsData} />
  </div>
);

export const WithFilter = () => (
  <div style={{ background: "white", width: "100%", margin: "auto" }}>
    <List {...(propsData as ListProps)} itemFilter="3" {...actionsData} />
  </div>
);
