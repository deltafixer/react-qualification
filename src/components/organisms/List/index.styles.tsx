import styled from "styled-components";

export const ListStyleWrapper = styled.ul`
  display: flex;
  flex-direction: column;
  margin: 0;
  padding: 0;

  li {
    &:last-of-type {
      margin-bottom: 0;
    }
  }
`;
