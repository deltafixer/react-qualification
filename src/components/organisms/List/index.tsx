import React, { useEffect } from "react";

type ListProps = {
  items: any[];
  itemFilter: string;
  resultCount: Function;
};

const List = ({ items, itemFilter, resultCount }: ListProps): JSX.Element => {
  let count = 0;

  useEffect(() => {
    resultCount(count);
  }, [count, itemFilter, resultCount]);

  return (
    <div className="items-container">
      {items.map((item, index) => {
        if (item.name.toLowerCase().includes(itemFilter.toLowerCase())) {
          count += 1;
          return <div key={index}>{item.name}</div>;
        }
        return null;
      })}
    </div>
  );
};

List.defaultProps = {
  text: "",
  resultCount: () => {},
};

export default List;
