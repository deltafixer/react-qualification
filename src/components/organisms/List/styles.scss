.items-container {
  display: flex;
  flex-direction: column;

  div {
    margin-bottom: 3px;

    &:last-of-type {
      margin-bottom: 0;
    }
  }
}
