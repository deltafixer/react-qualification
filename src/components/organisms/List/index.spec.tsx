import React from "react";
import { mount, ReactWrapper } from "enzyme";
import List, { ListProps } from "./index";

let wrapper: ReactWrapper;

jest.mock("../../molecules/ListItem", () => () => <div />);

const props: ListProps = {
  items: [
    {
      name: "name1",
      updated_at: "2020-01-01",
      created_at: "2020-01-01",
      owner: {
        avatar_url: "",
      },
    },
    {
      name: "name2",
      updated_at: "2020-01-03",
      created_at: "2020-01-01",
      owner: {
        avatar_url: "",
      },
    },
    {
      name: "name3",
      updated_at: "2020-01-02",
      created_at: "2020-01-01",
      owner: {
        avatar_url: "",
      },
    },
    {
      name: "value4",
      updated_at: "2020-01-02",
      created_at: "2020-01-01",
      owner: {
        avatar_url: "",
      },
    },
  ],
  itemFilter: "name",
  resultCount: jest.fn(),
};

describe("List component tests", () => {
  beforeAll(() => {
    wrapper = mount(<List {...props} />);
  });

  afterAll(() => {
    wrapper.unmount();
  });

  it("renders correctly", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
