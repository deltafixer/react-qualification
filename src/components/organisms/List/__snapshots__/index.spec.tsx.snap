// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component tests renders correctly 1`] = `
"<List items={{...}} itemFilter=\\"name\\" resultCount={[Function: mockConstructor]}>
  <styled.ul>
    <ul className=\\"sc-AxjAm bbeNOP\\">
      <Component name=\\"name2\\" avatar_url=\\"\\" created_at=\\"2020-01-01\\" updated_at=\\"2020-01-03\\">
        <div />
      </Component>
      <Component name=\\"name3\\" avatar_url=\\"\\" created_at=\\"2020-01-01\\" updated_at=\\"2020-01-02\\">
        <div />
      </Component>
      <Component name=\\"name1\\" avatar_url=\\"\\" created_at=\\"2020-01-01\\" updated_at=\\"2020-01-01\\">
        <div />
      </Component>
    </ul>
  </styled.ul>
</List>"
`;
